File: packages/python/client/starfish_sdk/kv_cache.py
"""KV-backed PullCache factory.

Mirrors ``packages/ts/client/src/kv-cache.ts``.

Adapts any :class:`KvStore` (async ``get``/``set``) into a :class:`PullCache`
that the ``StarfishClient`` can use as its offline read-through cache.

The cache stores a JSON envelope around each payload::

    {"payload": "<raw>", "_cached_at": <epoch_ms>}

so that optional max-age expiry can be checked on read without a separate
metadata key. Reading a plain-string entry (no ``_cached_at``) treats it as
fresh for backward-compatibility.

All ``get``/``set`` errors are swallowed — a failing KV store degrades to
"no cache" without crashing the caller.
"""

from __future__ import annotations

import json
import time
from typing import Optional, Protocol, runtime_checkable


# ── KvStore protocol ──────────────────────────────────────────────────────────

@runtime_checkable
class KvStore(Protocol):
    """Minimal async KV store interface."""

    async def get(self, key: str) -> Optional[str]:
        """Return the stored string value, or ``None`` if the key is absent."""
        ...

    async def set(self, key: str, value: str) -> None:
        """Persist *value* under *key*."""
        ...


# ── PullCache protocol ────────────────────────────────────────────────────────

@runtime_checkable
class PullCache(Protocol):
    """Async read/write cache for pull results (raw JSON strings)."""

    async def read(self, key: str) -> Optional[str]:
        """Return the cached raw payload string, or ``None`` on a miss/expiry."""
        ...

    async def write(self, key: str, data: str) -> None:
        """Persist *data* under *key*."""
        ...


# ── KvPullCache ───────────────────────────────────────────────────────────────

class KvPullCache:
    """PullCache backed by an async KV store.

    Stores JSON-serialised pull results in an envelope
    ``{"payload": "<raw>", "_cached_at": <epoch_ms>}``
    with optional max-age expiry.
    """

    def __init__(
        self,
        kv: KvStore,
        prefix: str = "starfish.pullcache.",
        max_age_ms: Optional[int] = None,
    ) -> None:
        self._kv = kv
        self._prefix = prefix
        self._max_age_ms = max_age_ms

    async def read(self, key: str) -> Optional[str]:
        """Return the cached payload, or ``None`` on miss / expiry / error."""
        try:
            raw = await self._kv.get(self._prefix + key)
            if raw is None:
                return None

            # Try the envelope format first; fall back to plain-string for
            # backward-compatibility with caches written before this library.
            payload: str
            cached_at: Optional[int] = None
            try:
                envelope = json.loads(raw)
                if isinstance(envelope, dict) and isinstance(envelope.get("payload"), str):
                    payload = envelope["payload"]
                    cached_at = envelope.get("_cached_at")
                else:
                    payload = raw
            except (json.JSONDecodeError, ValueError):
                payload = raw

            if self._max_age_ms is not None and cached_at is not None:
                age_ms = int(time.time() * 1000) - cached_at
                if age_ms > self._max_age_ms:
                    return None

            return payload
        except Exception:
            return None

    async def write(self, key: str, data: str) -> None:
        """Persist *data* under the prefixed *key*. Errors are silently swallowed."""
        try:
            cached_at = int(time.time() * 1000)
            envelope = json.dumps({"payload": data, "_cached_at": cached_at})
            await self._kv.set(self._prefix + key, envelope)
        except Exception:
            pass


# ── factory ───────────────────────────────────────────────────────────────────

def create_kv_pull_cache(
    kv: KvStore,
    prefix: str = "starfish.pullcache.",
    max_age_ms: Optional[int] = None,
) -> KvPullCache:
    """Wrap *kv* in a :class:`KvPullCache`.

    Args:
        kv:         Any async KV store (``get``/``set``).
        prefix:     Key prefix for cache entries. Change to avoid collisions
                    when the KV store is shared with other data.
        max_age_ms: Maximum age in milliseconds. Entries older than this are
                    treated as misses. ``None`` (default) means never expire.
    """
    return KvPullCache(kv, prefix=prefix, max_age_ms=max_age_ms)

File: packages/ts/client/src/kv-cache.ts
/**
 * KV-backed {@link PullCache} factory.
 *
 * {@link createKvPullCache} adapts any {@link AsyncStateStorage} (or any
 * object with `getItem`/`setItem`) into a {@link PullCache} that the
 * `StarfishClient` can use as its offline read-through cache.
 *
 * Why this matters: the client's `cache` option is ciphertext-at-rest by
 * construction — it stores the raw sealed server response and only decrypts
 * in memory on read. Backing the cache with the platform's own KV (AsyncStorage
 * on React Native, IndexedDB via `createIndexedDBStorage`, a custom adapter)
 * gives offline-first reads without exposing plaintext to the OS storage layer.
 *
 * @example
 * ```ts
 * import AsyncStorage from "@react-native-async-storage/async-storage"
 * import { StarfishClient } from "@drakkar.software/starfish-client"
 * import { createKvPullCache } from "@drakkar.software/starfish-client"
 *
 * const client = new StarfishClient({
 *   baseUrl: "https://api.example.com",
 *   cache: createKvPullCache(AsyncStorage, { prefix: "sf:", maxAgeMs: 30 * 24 * 60 * 60 * 1000 }),
 * })
 * ```
 */
import type { PullCache } from "./types.js"

/** A storage backend that `createKvPullCache` can wrap. */
export interface KvStore {
  getItem(key: string): Promise<string | null> | string | null
  setItem(key: string, value: string): Promise<unknown> | unknown
  removeItem?: (key: string) => Promise<unknown> | unknown
}

/** Options for {@link createKvPullCache}. */
export interface KvPullCacheOptions {
  /**
   * Key prefix for cache entries (default `"starfish.pullcache."`). Change
   * when sharing a KV store with other data to avoid key collisions.
   */
  prefix?: string
  /**
   * Maximum age in milliseconds for a cached snapshot. When set, an entry
   * older than `maxAgeMs` is returned as `null` (a cache MISS) so the next
   * pull goes to the network rather than serving arbitrarily stale data.
   *
   * Each entry stores a `_cachedAt` wall-clock timestamp; expiry is checked on
   * every `get`. Omit (default) for entries that never expire — recommended
   * for offline-first apps where any last-synced data beats none.
   */
  maxAgeMs?: number
}

interface CacheEntry {
  payload: string
  _cachedAt: number
}

/**
 * Adapt a KV store into a {@link PullCache} for `StarfishClient`.
 *
 * The adapter serialises the cached pull payload (itself a JSON string) into
 * an outer JSON envelope that tracks the wall-clock write time for optional
 * max-age expiry. Reading a legacy entry without `_cachedAt` treats it as
 * fresh (backward-compatible with plain-string caches).
 *
 * All `get`/`set` errors are swallowed (the {@link PullCache} contract
 * requires implementations not to throw) — a failing KV store simply
 * degrades to "no cache" without crashing the app.
 */
export function createKvPullCache(kv: KvStore, opts: KvPullCacheOptions = {}): PullCache {
  const prefix = opts.prefix ?? "starfish.pullcache."
  const maxAgeMs = opts.maxAgeMs

  return {
    async get(key: string): Promise<string | null> {
      try {
        const raw = await kv.getItem(prefix + key)
        if (raw === null || raw === undefined) return null

        // Try to read the new envelope format `{"payload":"…","_cachedAt":n}`.
        // Fall back to treating the raw string as the payload directly for
        // backward-compatibility with plain-string caches written before this
        // library was used.
        let payload: string
        let cachedAt: number | undefined
        try {
          const envelope = JSON.parse(raw) as Partial<CacheEntry>
          if (typeof envelope.payload === "string") {
            payload = envelope.payload
            cachedAt = envelope._cachedAt
          } else {
            // Plain-string format: the raw value is the payload.
            payload = raw
          }
        } catch {
          payload = raw
        }

        if (maxAgeMs !== undefined && cachedAt !== undefined) {
          if (Date.now() - cachedAt > maxAgeMs) return null
        }

        return payload
      } catch {
        return null
      }
    },

    async set(key: string, value: string): Promise<void> {
      try {
        const envelope: CacheEntry = { payload: value, _cachedAt: Date.now() }
        await kv.setItem(prefix + key, JSON.stringify(envelope))
      } catch {
        // Swallow — storage failures degrade to "no cache", not a crash.
      }
    },
  }
}
